File: app/src/main/java/edu/upc/mcia/publications/ui/authors/AuthorsFragment.java
package edu.upc.mcia.publications.ui.authors;


import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.view.MenuItemCompat;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.jakewharton.rxbinding.support.v4.view.RxMenuItemCompat;
import com.jakewharton.rxbinding.support.v7.widget.RxSearchView;
import com.jakewharton.rxbinding.support.v7.widget.SearchViewQueryTextEvent;
import com.jakewharton.rxbinding.view.MenuItemActionViewEvent;

import java.util.List;

import edu.upc.mcia.publications.R;
import edu.upc.mcia.publications.data.model.Author;
import timber.log.Timber;

/**
 * A simple {@link Fragment} subclass.
 */
public class AuthorsFragment extends Fragment implements AuthorsMvpView {

    private AuthorsPresenter mPresenter;

    private RecyclerView mRecyclerView;
    private RecyclerView.LayoutManager mLayoutManager;
    private AuthorsAdapter mAdapter;

    public AuthorsFragment() {
        // Required empty public constructor
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        mPresenter = new AuthorsPresenter();
        setHasOptionsMenu(true);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_authors, container, false);
        mRecyclerView = (RecyclerView) view.findViewById(R.id.recyclerView);

        mLayoutManager = new LinearLayoutManager(getContext());
        mAdapter = new AuthorsAdapter();
        mAdapter.setOnItemClickListener((v, author) -> Timber.d("Item clicked"));

        mRecyclerView.setLayoutManager(mLayoutManager);
        mRecyclerView.setHasFixedSize(true);
        mRecyclerView.setAdapter(mAdapter);

        return view;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mPresenter.attachView(this);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.menu_authors, menu);
        MenuItem searchItem = menu.findItem(R.id.action_search);
        SearchView searchView = (SearchView) MenuItemCompat.getActionView(searchItem);

        RxMenuItemCompat.actionViewEvents(searchItem)
                .filter(event -> event.kind().equals(MenuItemActionViewEvent.Kind.COLLAPSE))
                .subscribe(event -> mPresenter.onSearchQueryDismissed());

        RxSearchView.queryTextChangeEvents(searchView)
                .filter(SearchViewQueryTextEvent::isSubmitted)
                .map(query -> query.queryText().toString().trim())
                .subscribe(query -> mPresenter.onSearchQuerySubmitted(query));

        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mPresenter.detachView();
    }

    @Override
    public void showAuthors(List<Author> authors) {
        Timber.d("Received authors");
        mAdapter.setData(authors);
    }

    @Override
    public void filterAuthors(String query) {
        mAdapter.getFilter().filter(query);
    }
}

File: app/src/main/java/edu/upc/mcia/publications/ui/authors/AuthorsFragment.kt
package edu.upc.mcia.publications.ui.authors

import android.arch.lifecycle.Observer
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.View
import com.jakewharton.rxbinding2.support.v7.widget.queryTextChangeEvents
import edu.upc.mcia.publications.BR
import edu.upc.mcia.publications.R
import edu.upc.mcia.publications.data.model.Author
import edu.upc.mcia.publications.databinding.FragmentAuthorsBinding
import edu.upc.mcia.publications.ui.base.BaseFragment
import edu.upc.mcia.publications.ui.base.BaseItem
import edu.upc.mcia.publications.ui.base.BaseItemBinding
import edu.upc.mcia.publications.ui.base.BaseItemDiffCallback
import edu.upc.mcia.publications.ui.common.ErrorItem
import edu.upc.mcia.publications.ui.common.LoadingItem
import edu.upc.mcia.publications.ui.utils.bindingProvider
import edu.upc.mcia.publications.ui.utils.viewModelProvider
import me.tatarka.bindingcollectionadapter2.OnItemBind
import me.tatarka.bindingcollectionadapter2.collections.DiffObservableList

/**
 * A simple [Fragment] subclass.
 */
class AuthorsFragment : BaseFragment() {

    override val binding: FragmentAuthorsBinding by bindingProvider(R.layout.fragment_authors, BR.authorsFragment to this)
    private val model by viewModelProvider { AuthorsViewModel() }

    val itemBinding: OnItemBind<BaseItem> = BaseItemBinding()
    val items: DiffObservableList<BaseItem> = DiffObservableList(BaseItemDiffCallback())

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.searchView.queryTextChangeEvents()
                .map { it.queryText().toString().trim { it <= ' ' } }
                .startWith("")
                .subscribe { model.uiEvents.accept(AuthorsViewModel.UiEvent.QueryChange(it)) }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        model.uiState.observe(this, Observer {
            if (it!!.loading) items.update(listOf(LoadingItem()))
            if (it.error != null) items.update(listOf(ErrorItem(it.error.message, { model.uiEvents.accept(it.error.trigger) })))
            showAuthors(it.authors)
        })
    }

    private fun showAuthors(authors: List<Author>) {
        val listItems = mutableListOf<BaseItem>()

        listItems += authors
                .filter { it.status == "member" }
                .map { AuthorItem(it) }
                .sortedBy { it.sortingName }
                .also { if (it.isNotEmpty()) listItems += AuthorHeaderItem("Members") }

        listItems += authors
                .filter { it.status == "external" }
                .map { AuthorItem(author = it, showAvatar = false) }
                .sortedBy { it.sortingName }
                .also { if (it.isNotEmpty()) listItems += AuthorHeaderItem("External") }

        if (listItems.isEmpty()) listItems += AuthorEmptyItem()

        items.update(listItems)
        binding.recyclerView.scrollToPosition(0)
    }
}
